### PR TITLE
feat(coding-graph): Phase B type resolution — LSP client layer (#1555)

### PR DESCRIPTION
**Pull request by joshuaswarren**
## Summary

Implements issue #1555 — Phase B type resolution via a minimal JSON-RPC-over-stdio LSP client layer over installed language servers. Drives real language servers (typescript-language-server, pyright, gopls, rust-analyzer) as subprocesses and uses their `textDocument/definition` responses to upgrade Phase A heuristic edges with resolved-type information.

Part of #1548 (Track B, Phase 3). Depends on #1551 (parser) + #1552 (store) — both merged.

### LSP Client (`lsp/client.ts`)
- Minimal JSON-RPC-over-stdio client — no npm LSP framework dependency
- Lifecycle: `initialize` → `initialized` → `didOpen` → `definition` → `shutdown` → `exit`
- Request/response correlation by id with per-request timeout
- Zombie cleanup: `dispose()` sends shutdown+exit, then SIGKILL — no child process survives (tested via pid liveness)
- All failure modes degrade to distinct `LspDegradation` codes — never throws

### Frame Decoder (`lsp/framing.ts`)
- Byte-accurate `Content-Length` frame decoder (Buffer-based)
- Handles UTF-8 multi-byte bodies, split-buffer chunks, scan-offset optimization (rule 32)
- Distinct decode errors: `malformed_header`, `json_parse_error`

### Config (`lsp/config.ts`)
- `codingGraph.lsp.enabled` defaults `false` (rule 30/48)
- Env override: `REMNIC_CODING_GRAPH_LSP_ENABLED` with `ENGRAM_` fallback (gotcha 9)
- Server overrides validation: unknown language → error listing supported (rule 51)

### Resolution Pass (`lsp/resolution.ts`)
- **Pure planner**: `planLspUpgrades(unresolvedCallSites, budget)` → requests with deterministic ordering (rule 38) and budget enforcement
- **Executor**: sends definition requests, maps LSP locations to graph nodes by half-open span containment (rule 35), applies edge upgrades transactionally per file batch (rule 25)
- Mid-batch `applyUpgrades` failure → caught, zero partial upgrades persist
- Server crash mid-run → degradation surfaced, remaining sites keep Phase A results

### Status Surfacing (`lsp/status.ts`)
- Per-language state: `enabled/probed/degraded(code)/requests_used`
- `formatLspStatusLine` for `remnic doctor` rendering

### Byte-Position Converter (`lsp/byte-position.ts`)
- `buildLineOffsetMap` + `byteOffsetToPosition` + `positionToByteOffset`
- Handles `\n`, `\r\n`, `\r` line endings and UTF-8 multi-byte chars

## Test Coverage

**46 new tests, all green:**
- Framing: 11 edge-case tests (split body/header, multi-frame, UTF-8 bytes vs code units, malformed header, invalid JSON, scan-offset reset)
- Client: 10 tests — **5-mode fake-server matrix** (happy/server_missing/handshake_timeout/request_timeout/protocol_error), zombie cleanup (pid liveness), crash_after_start, dispose idempotency, definition after dispose
- Resolution: 15 tests (planner purity/budget/ordering, location mapping, executor happy/unresolved/mid-batch-fail/crash/timeout/empty)
- Characterization: 10 tests (default disabled, enabled+zero servers → Phase A identical, status rendering, degradation codes, empty-request safety)

**Full suite: 318 pass / 0 fail** (272 original + 46 new).

## Done-When Checklist

- [x] Fake-server matrix (5 modes) green
- [x] Framing edge-case tests green
- [x] Zombie-cleanup test green (pid liveness)
- [x] Mid-batch abort leaves no partial upgrades
- [x] With `lsp.enabled: false` the index output is byte-identical to Phase A (characterization)
- [x] Doctor/index_status lines render
- [x] Ratchets untouched (coding-graph is outside the ratchet watchlist)
- [x] `tsc --noEmit` clean
- [x] `pnpm test` — 318 pass / 0 fail
- [ ] Real-server integration test (behind `REMNIC_TEST_LSP=1` env opt-in) — optional per issue; the fake-server matrix covers all protocol paths

## Chokepoints used / not applicable
- Not applicable: this PR adds a new `lsp/` directory to `@remnic/coding-graph` (no core file edits, no orchestrator/storage/access/config changes)

---

> [!NOTE]
> **Medium Risk**
> Spawns child language-server processes and mutates graph edges when enabled; misconfiguration or partial failures are mitigated by degradations and default-off, but integration with the index pipeline still needs careful gating and transactional apply.
> 
> **Overview**
> Introduces **Phase B type resolution** (#1555): a new `packages/coding-graph/src/lsp/` stack and public exports from `@remnic/coding-graph` so callers can optionally drive installed language servers and refine heuristic **CALLS** edges. Everything is gated by **`codingGraph.lsp.enabled` (default `false`)**; characterization tests expect the index pipeline to skip LSP when disabled.
> 
> **Client & transport:** Custom **LSP 3.17** subset over stdio—`Content-Length` framing (`framing.ts`), minimal **`LspClient`** (initialize → didOpen → definition, shutdown/exit + SIGKILL on dispose), and **`pathToUri` / `uriToPath`**. Failures return tagged **`LspDegradation`** codes (never throw).
> 
> **Resolution:** **`planLspUpgrades`** (budget, deterministic ordering) and **`executeLspResolution`** (per-file batches, didOpen before definition, map locations to graph nodes via byte spans, apply **`provenance: "lsp"`** upgrades with per-batch atomicity on crash/apply failure).
> 
> **Config & ops:** **`parseLspConfig`**, env **`REMNIC_CODING_GRAPH_LSP_ENABLED`** / **`ENGRAM_`** fallback, **`getLspStatus`** / **`formatLspStatusLine`** for doctor/index status. **`byte-position.ts`** bridges UTF-8 store spans and LSP UTF-16 positions.
> 
> **Tests:** Fake-server matrix, framing edge cases, resolution/characterization suites (~46 new tests); no core index wiring appears in this diff—the layer is exported for integration behind the enabled flag.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 8806f0cd346f6c6e310e648c7cd0f14ccc3f57c4. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/coding-graph/src/lsp/byte-position.ts [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+ */
+export function buildLineOffsetMap(content: string | Buffer): LineOffsetMap {
+  const text = typeof content === "string" ? content : content.toString("utf8");
+  const lineStarts = [0];
+  // Scan for line endings. `\r\n` is counted as one break (don't double-count).
+  for (let i = 0; i < text.length; i++) {
+    const ch = text.charCodeAt(i);
+    if (ch === 0x0a) {
+      // \n — Unix line ending (or the \n of a \r\n pair).
+      lineStarts.push(i + 1);
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/lsp/byte-position.ts]: Fixed in 89121bfe — see commit for details.

File: packages/coding-graph/src/lsp/byte-position.ts [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    } else if (ch === 0x0d) {
+      // \r — could be \r\n or lone \r.
+      if (i + 1 < text.length && text.charCodeAt(i + 1) === 0x0a) {
+        // \r\n — skip the \n in the next iteration by advancing i.
+        i++;
+      }
+      lineStarts.push(i + 1);
+    }
+  }
+  return { lineStarts };
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/lsp/byte-position.ts:70]: Fixed in 89121bfe — see commit for details.

File: packages/coding-graph/src/lsp/byte-position.ts [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+      i++; // skip the low surrogate
+    } else if (code < 0x80) {
+      currentByte += 1;
+    } else if (code < 0x800) {
+      currentByte += 2;
+    } else {
+      currentByte += 3;
+    }
+    charIndex++;
+  }
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/lsp/byte-position.ts:111]: Fixed in 89121bfe — see commit for details.

File: packages/coding-graph/src/lsp/resolution.ts [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+ * via options).
+ */
+function filePathToUri(filePath: string): string {
+  // Simple file:// URI for repo-relative paths. The LSP server resolves
+  // relative to rootUri. For absolute paths, the path is used as-is.
+  const isAbsolute = filePath.startsWith("/") || /^[A-Za-z]:[\\/]/.test(filePath);
+  if (isAbsolute) {
+    return `file://${filePath.replace(/\\/g, "/")}`;
+  }
+  return `file:///${filePath.replace(/\\/g, "/")}`;
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/lsp/resolution.ts:465]: Fixed in 89121bfe — see commit for details.

[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/lsp/resolution.ts:465]: Fixed in 89121bfe — see commit for details.

File: packages/coding-graph/src/lsp/resolution.ts [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+ *
+ * Returns null if no location maps to an indexed node.
+ */
+export function mapLocationToNode(
+  locations: readonly LspLocation[],
+  callerContent: string,
+  nodeLocator: NodeLocator,
+): string | null {
+  for (const loc of locations) {
+    const filePath = uriToPath(loc.uri);
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/lsp/resolution.ts]: Fixed in 89121bfe — see commit for details.

File: packages/coding-graph/src/lsp/resolution.ts [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+    // files' node spans.
+    //
+    // For the byte conversion, we pass the caller's content as a
+    // best-effort. If the definition is in the same file, this is exact.
+    // If it's in a different file, the line/character → byte conversion
+    // may be slightly off for files with different encodings. The
+    // nodeLocator uses half-open containment which tolerates small
+    // offsets (the byte offset just needs to fall within the node's span).
+    const map = buildLineOffsetMap(callerContent);
+    const startByte = positionToByteOffset(callerContent, loc.range.start, map);
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/lsp/resolution.ts]: Fixed in 89121bfe — see commit for details.

File: packages/coding-graph/src/lsp/resolution.ts [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+    const qName = nodeLocator(filePath, startByte);
+    if (qName !== null) return qName;
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/lsp/resolution.ts:439]: Fixed in 89121bfe — see commit for details.

File: packages/coding-graph/src/lsp/resolution.ts [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+
+      if (!defResult.ok) {
+        // Distinguish "server problem" (stop the whole pass) from
+        // "this particular definition returned nothing" (continue).
+        if (
+          defResult.degradation.code === "server_crashed" ||
+          defResult.degradation.code === "protocol_error"
+        ) {
+          degradation = defResult.degradation;
+          break;
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/lsp/resolution.ts:328]: Fixed in 89121bfe — see commit for details.

File: packages/coding-graph/src/lsp/resolution.ts [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+    const upgrades: EdgeUpgrade[] = [];
+    let batchFailed = false;
+
+    for (const req of batchReqs) {
+      if (degradation) break; // server is dead — stop sending
+
+      const defResult = await client.definition({
+        textDocument: { uri: filePathToUri(req.filePath) },
+        position: req.position,
+      });
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/lsp/resolution.ts:311]: Fixed in 89121bfe — see commit for details.

File: packages/coding-graph/src/lsp/config.ts [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+        backend: "lsp",
+        code: "protocol_error",
+        detail: "codingGraph.lsp must be an object",
+      },
+    };
+  }
+  const obj = raw as Record<string, unknown>;
+  const knownSet = new Set(knownLanguages);
+
+  const enabled = obj.enabled === undefined ? false : Boolean(obj.enabled);
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/lsp/config.ts]: Fixed in 89121bfe — see commit for details.

File: packages/coding-graph/src/lsp/client.ts [hunk 11]
```diff
@@ -0,0 +1,461 @@
...
+          ? "handshake_timeout"
+          : initResult.degradation.code;
+      return {
+        ok: false,
+        degradation: lspDegradation(code, initResult.degradation.detail),
+      };
+    }
+
+    const initResponse = initResult.value as LspInitializeResult;
+    client.serverCapabilities = initResponse.capabilities;
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/lsp/client.ts:179]: Fixed in bd106e3b — see commit for details.

File: packages/coding-graph/src/lsp/client.ts [hunk 12]
```diff
@@ -0,0 +1,461 @@
...
+  /**
+   * Handle a spawn error (ENOENT etc). Marks the server as missing.
+   */
+  private onChildError(err: Error): void {
+    // ENOENT = binary not found.
+    this.crashed = true;
+    const detail = err.message.includes("ENOENT") ? undefined : err.message;
+    for (const [id, entry] of this.pending) {
+      clearTimeout(entry.timer);
+      entry.reject(lspDegradation("server_missing", detail));
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/lsp/client.ts]: Fixed in bd106e3b — see commit for details.

File: packages/coding-graph/src/lsp/client.ts [hunk 13]
```diff
@@ -0,0 +1,461 @@
...
+   */
+  static async connect(options: LspClientOptions): Promise<
+    | { ok: true; client: LspClient }
+    | { ok: false; degradation: LspDegradation }
+  > {
+    let child: ChildProcess;
+    const spawnFn = options.spawnFn ?? spawn;
+    try {
+      child = spawnFn(options.launchSpec.command, [...options.launchSpec.args], {
+        stdio: ["pipe", "pipe", "pipe"],
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/lsp/client.ts:113]: Fixed in 8e0782e9 — the constructor now attaches an empty reader to child.stderr to drain the pipe buffer.

File: packages/coding-graph/src/lsp/client.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/lsp/client.ts:179]: Fixed in bd106e3b — connect() now checks that initResult.value is a non-null object before accessing .capabilities.

File: packages/coding-graph/src/lsp/client.ts [hunk 14]
```diff
@@ -0,0 +1,461 @@
...
+ */
+export function uriToPath(uri: string): string {
+  if (uri.startsWith("file://")) {
+    const rest = uri.slice("file://".length);
+    // Windows: file:///C:/foo → C:/foo
+    if (/^\/[A-Za-z]:/.test(rest)) {
+      return rest.slice(1).replace(/\//g, path.sep);
+    }
+    // Unix: file:///foo → /foo
+    return rest.replace(/\//g, path.sep);
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/lsp/client.ts:481]: Fixed in 8e0782e9 — uriToPath now calls decodeURIComponent to handle percent-encoded URIs per RFC 8089.

File: packages/coding-graph/src/lsp/client.ts [hunk 15]
```diff
@@ -0,0 +1,468 @@
...
+// ──────────────────────────────────────────────────────────────────────────
+
+/**
+ * Convert a repo-relative or absolute file path to a `file://` URI.
+ * Handles Windows drive letters (C:\ → file:///C:/).
+ */
+export function pathToUri(filePath: string): string {
+  const abs = path.isAbsolute(filePath) ? filePath : path.resolve(filePath);
+  // On Windows, prepend / before the drive letter.
+  return `file://${abs.replace(/\\/g, "/")}`;
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/lsp/client.ts]: Fixed in 0a377c71 — pathToUri now prepends a leading slash for Windows drive paths (C:/foo → file:///C:/foo) per RFC 8089.

File: packages/coding-graph/src/lsp/client.ts [hunk 16]
```diff
@@ -0,0 +1,468 @@
...
+
+    const client = new LspClient(child, options.rootUri, options.timeoutMs);
+
+    // If the spawn errored asynchronously (ENOENT fires as 'error' event),
+    // the client.crashed flag is set by onChildError. Check after wiring.
+    if (client.crashed) {
+      return {
+        ok: false,
+        degradation: lspDegradation("server_missing"),
+      };
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/lsp/client.ts:145]: Fixed in 0a377c71 — non-ENOENT spawn errors now report server_crashed; onChildExit sets crashCode=server_crashed; connect() uses crashCode instead of hardcoding server_missing.

File: packages/coding-graph/src/lsp/config.ts [hunk 17]
```diff
@@ -0,0 +1,220 @@
...
+  maxRequestsPerRun: DEFAULT_LSP_MAX_REQUESTS_PER_RUN,
+};
+
+/**
+ * Parse and validate user-supplied LSP config. Unknown keys in `servers`
+ * are rejected with a degradation listing supported languages (rule 51).
+ * Non-executable absolute paths are rejected (rule 24 analog).
+ *
+ * Returns `{ ok: true, config }` or `{ ok: false, degradation }` — never
+ * throws (rule 13).
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/lsp/config.ts:70]: Fixed in 0a377c71 — removed the dead spawn import from config.ts.

File: packages/coding-graph/src/lsp/resolution.ts [hunk 18]
```diff
@@ -0,0 +1,476 @@
...
+  /**
+   * Apply a batch of edge upgrades atomically. Called once per file batch.
+   * MUST be transactional — if it throws, zero upgrades from this batch
+   * persist (rule 25). Each upgrade is an edge `{srcQualifiedName,
+   * dstQualifiedName, type: "CALLS", confidence, provenance: "lsp"}`.
+   */
+  readonly applyUpgrades: (
+    upgrades: readonly EdgeUpgrade[],
+  ) => Promise<void>;
+  readonly perRequestTimeoutMs?: number;
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/lsp/resolution.ts]: Fixed in 0a377c71 — removed the unused perRequestTimeoutMs field from ResolveOptions. The LspClient's own timeoutMs is sufficient.

File: packages/coding-graph/src/lsp/resolution.ts [hunk 19]
```diff
@@ -0,0 +1,475 @@
...
+        if (
+          defResult.degradation.code === "server_crashed" ||
+          defResult.degradation.code === "protocol_error"
+        ) {
+          degradation = defResult.degradation;
+          // Mark the batch failed so collected upgrades are NOT committed
+          // (rule 25 — per-batch atomicity; a degraded batch must not
+          // persist partial LSP edges).
+          batchFailed = true;
+          break;
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/lsp/resolution.ts:328]: Fixed in 8806f0cd — fatal errors now increment unresolved for the failed request and all remaining unattempted requests in the batch, matching the request_timeout/request_error accounting.
